### PR TITLE
[GraphOptimizer] Fix batchNorm optimization

### PR DESCRIPTION
**Pull request by qcolombet**
The batchNorm optimization used to assume that the inputs of the related
convolution node were all variables. This is not enforced anywhere and may
actually be wrong.

The fix consists in checking that the inputs of the convolution node are indeed
variables. Moreover, we check that these variables are used only by this
convolution node before doing the transformation. Otherwise we will incorrectly
change the inputs of the other users.

This fixes issue #1178 for the graph optimizer.

### CODE REVIEW

**Comment by qcolombet** [comment]
Note: This pull request depends on a fix for PR #1184.
Otherwise the batchNorm optimization doesn't kick in at all because getFilter creates a temporary NodeValue that increases the #users by 1.

File: lib/Optimizer/GraphOptimizer.cpp [hunk 1]
```diff
@@ -763,6 +763,18 @@ static void optimizePool(Function *F) {
...
+static Variable *getModifiableVariable(Node &node) {
+  // If that node has more than one use, it may not
+  // be okay to modify the underlying variable.
+  if (!node.hasOneUse()) {
+    return nullptr;
+  }
+  if (!isa<Variable>(&node)) {
+    return nullptr;
+  }
+  return cast<Variable>(&node);
```
Comment (bertmaher) [inline lib/Optimizer/GraphOptimizer.cpp]: Could replace the isa/cast pair with dyn_cast.  Up to you though, it's pretty clear this way.

File: lib/Optimizer/GraphOptimizer.cpp [hunk 2]
```diff
@@ -815,13 +827,24 @@ static void optimizeBatchNorm(Function *F) {
...
 
-      auto filterH = cast<Variable>(CV->getFilter())->getHandle<>();
-      auto cbiasH = cast<Variable>(CV->getBias())->getHandle<>();
+      Variable *filterV = getModifiableVariable(*CV->getFilter());
+      Variable *cbiasV = getModifiableVariable(*CV->getBias());
+
+      Variable *scaleV = getModifiableVariable(*BN->getScale());
+      Variable *biasV = getModifiableVariable(*BN->getBias());
+      Variable *meanV = getModifiableVariable(*BN->getMean());
+      Variable *var = getModifiableVariable(*BN->getVar());
```
Comment (bertmaher) [inline lib/Optimizer/GraphOptimizer.cpp]: The BatchNorm variables don't technically have to be single-use, right?  Since we're actually modifying the Conv inputs?

Comment (qcolombet) [reply lib/Optimizer/GraphOptimizer.cpp]: Good catch!

File: lib/Optimizer/GraphOptimizer.cpp [hunk 3]
```diff
@@ -763,6 +763,15 @@ static void optimizePool(Function *F) {
   } // For all nodes in the graph.
 }
 
+static Variable *getModifiableVariable(Node &node) {
```
Comment (nadavrot) [inline lib/Optimizer/GraphOptimizer.cpp]: Doxygen comments please.

Also, the name is not great, because "Modifiable" is subjective and depends on the context.

**Comment by nadavrot** [comment]
@qcolombet To remove the dependency on the NodeValue refactoring you could use getFilter().getNode().

File: lib/Optimizer/GraphOptimizer.cpp [hunk 4]
```diff
@@ -815,13 +824,24 @@ static void optimizeBatchNorm(Function *F) {
...
 
-      auto filterH = cast<Variable>(CV->getFilter())->getHandle<>();
-      auto cbiasH = cast<Variable>(CV->getBias())->getHandle<>();
+      Variable *filterV = getModifiableVariable(*CV->getFilter());
+      Variable *cbiasV = getModifiableVariable(*CV->getBias());
+
+      Variable *scaleV = dyn_cast<Variable>(BN->getScale());
+      Variable *biasV = dyn_cast<Variable>(BN->getBias());
+      Variable *meanV = dyn_cast<Variable>(BN->getMean());
+      Variable *var = dyn_cast<Variable>(BN->getVar());
```
Comment (rdzhabarov) [inline lib/Optimizer/GraphOptimizer.cpp]: based on the way we build batch normalization node we expect all those 4 be a variable, maybe place an assert on those instead of if checking.
This way if it changes we could adopt implementation of optimization otherwise this optimization will not trigger at all (would be easy to skip).

File: lib/Optimizer/GraphOptimizer.cpp [hunk 5]
```diff
@@ -815,13 +824,24 @@ static void optimizeBatchNorm(Function *F) {
       // Q = W * A
       // C = b * A + B
 
-      auto filterH = cast<Variable>(CV->getFilter())->getHandle<>();
-      auto cbiasH = cast<Variable>(CV->getBias())->getHandle<>();
+      Variable *filterV = getModifiableVariable(*CV->getFilter());
```
Comment (rdzhabarov) [inline lib/Optimizer/GraphOptimizer.cpp]: can you also update comment above:
```
 // Merge the Batch Normalization operation into the convolution that comes
     // before it by updating the weights of the filter.
```
It changes convolution filter and bias.

File: lib/Optimizer/GraphOptimizer.cpp (hunk 4 above)
Comment (qcolombet) [reply lib/Optimizer/GraphOptimizer.cpp]: That's the thing, this is going to change, because only load and save would be able to work directly with variables.

File: lib/Optimizer/GraphOptimizer.cpp (hunk 5 above)
Comment (qcolombet) [reply lib/Optimizer/GraphOptimizer.cpp]: Good catch!

**Comment by qcolombet** [comment]
Hi @nadavrot,

Thanks for the review.

One comment:
> @qcolombet To remove the dependency on the NodeValue refactoring you could use getFilter().getNode().

I've tried, but that doesn't work either. We still have a temporary NodeValue alive at the call site.
To really workaround this limitation, we would need to use a temporary variable like so:
```
Node *tmp = CV->getFilter().getNode();
Variable *filterV = getModifiableVariable(*tmp);
```

I can go with that to unblock this, but it feels unnatural.

File: lib/Optimizer/GraphOptimizer.cpp (hunk 4 above)
Comment (qcolombet) [reply lib/Optimizer/GraphOptimizer.cpp]: I'll go back to the cast, because I'll have to change this code anyway when the change for loads land. It used to be consistent when I was using getModifiableVariable for all, but now that I rolled back to a straight dyn_cast, that doesn't make much sense.

Anyway, reverting to the straight cast here.

File: lib/Optimizer/GraphOptimizer.cpp [hunk 6]
```diff
@@ -815,13 +826,33 @@ static void optimizeBatchNorm(Function *F) {
...
       // C = b * A + B
 
-      auto filterH = cast<Variable>(CV->getFilter())->getHandle<>();
-      auto cbiasH = cast<Variable>(CV->getBias())->getHandle<>();
+      // FIXME: We need to use a temporary node to hold on
+      // the variables, because using the result directly
+      // will create a temporary NodeValue that is going
+      // to be otherwise alive at the call site and that
+      // messes up with the number of users.
+      Node *tmp = CV->getFilter().getNode();
```
Comment (nadavrot) [inline lib/Optimizer/GraphOptimizer.cpp]: I wonder why getFilter returns NodeValue instead of NodeValue&. But that's a different story.

Comment (qcolombet) [reply lib/Optimizer/GraphOptimizer.cpp]: Agreed :).

**Comment by nadavrot** [comment]
Nice!
